File: node/src/auth/middleware.ts
import type { NextFunction, Request, Response } from "express";
import { pool } from "../db/pool.js";
import { hashApiKey } from "../lib/crypto.js";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Express {
    interface Request {
      tenantId: string;
    }
  }
}

// Bearer API key per tenant, resolved server-side to tenant_id — never
// accepted as a URL/body parameter (Shared REST API contract ticket).
export async function requireTenant(req: Request, res: Response, next: NextFunction): Promise<void> {
  try {
    const header = req.header("authorization") ?? "";
    const [scheme, token] = header.split(" ");
    if (scheme !== "Bearer" || !token) {
      res.status(401).json({ error: { code: "unauthorized", message: "Missing or malformed Authorization header" } });
      return;
    }

    const { rows } = await pool.query<{ id: string }>(
      "SELECT id FROM tenants WHERE api_key_hash = $1",
      [hashApiKey(token)],
    );
    const tenant = rows[0];
    if (!tenant) {
      res.status(401).json({ error: { code: "unauthorized", message: "Invalid API key" } });
      return;
    }

    req.tenantId = tenant.id;
    next();
  } catch (err) {
    next(err);
  }
}

File: go/internal/auth/middleware.go
// Package auth resolves a Bearer API key to a tenant_id, mirroring
// node/src/auth/middleware.ts.
package auth

import (
	"context"
	"net/http"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"

	"webhooks-go/internal/crypto"
	"webhooks-go/internal/httpx"
)

type contextKey int

const tenantIDKey contextKey = iota

// RequireTenant resolves the Bearer API key server-side to a tenant_id —
// never accepted as a URL/body parameter (Shared REST API contract).
func RequireTenant(pool *pgxpool.Pool) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			scheme, token, ok := strings.Cut(r.Header.Get("Authorization"), " ")
			if !ok || scheme != "Bearer" || token == "" {
				httpx.WriteError(w, http.StatusUnauthorized, "unauthorized", "Missing or malformed Authorization header")
				return
			}

			var tenantID string
			err := pool.QueryRow(r.Context(), "SELECT id FROM tenants WHERE api_key_hash = $1", crypto.HashAPIKey(token)).Scan(&tenantID)
			if err != nil {
				httpx.WriteError(w, http.StatusUnauthorized, "unauthorized", "Invalid API key")
				return
			}

			next.ServeHTTP(w, r.WithContext(context.WithValue(r.Context(), tenantIDKey, tenantID)))
		})
	}
}

// TenantID reads the tenant_id set by RequireTenant. Only ever called from
// inside a handler mounted behind that middleware, so an empty return here
// would indicate a routing bug, not a real unauthenticated request.
func TenantID(r *http.Request) string {
	id, _ := r.Context().Value(tenantIDKey).(string)
	return id
}
